aggregate/billing_functions/utils.py

```diff
@@ -0,0 +1,907 @@
+
+    # Process chunks of batches to avoid loading too many entries into memory
+    for btch_grp in chunk(batches, entry_chunk_size):
+        jobs_in_batch = []
+        entries = []
+
+        # Get jobs for a fraction of each chunked batches
+        # to avoid hitting hail batch too much
+        for batch_group_group in chunk(btch_grp, batch_group_chunk_size):
+            chnk_counter += 1
+            times = [b['time_created'] for b in batch_group_group]
+            min_batch = min(times)
+            max_batch = max(times)
+
+            if len(batches) > 100:
+                logger.debug(
+                    f'{lp}Getting jobs for batch chunk {chnk_counter}/{nchnks} '
+                    f'[{min_batch}, {max_batch}]'
+                )
+
+            promises = [get_jobs_for_batch(b['id'], token) for b in batch_group_group]
+            jobs_in_batch.extend(await asyncio.gather(*promises))
+
+        # insert all entries for this batch
+        for batch, jobs in zip(btch_grp, jobs_in_batch):
+            batch['jobs'] = jobs
+            if len(jobs) > 10000 and len(entries) > 1000:
+                logger.info(
+                    f'Expecting large number of jobs ({len(jobs)}) from '
+                    f'batch {batch["id"]}, inserting contents early'
+                )
+                result += insert_new_rows_in_table(
+                    table=GCP_AGGREGATE_DEST_TABLE, objs=entries, dry_run=dry_run
+                )
+                entries = []
+
+            entries_for_batch = func_get_finalised_entries_for_batch(batch)
+            entries.extend(entries_for_batch)
+
+            s = sum(sys.getsizeof(e) for e in entries) / 1024 / 1024
+            if s > 10:
+                logger.info(f'Size of entries: {s} MB, inserting early')
+                result += insert_new_rows_in_table(
+                    table=GCP_AGGREGATE_DEST_TABLE, objs=entries, dry_run=dry_run
+                )
+                entries = []
+
+        result += insert_new_rows_in_table(
+            table=GCP_AGGREGATE_DEST_TABLE, objs=entries, dry_run=dry_run
+        )
+
+    return result
+
+
+RE_matcher = re.compile(r'-\d+$')
+
+
+def billing_row_to_topic(row, dataset_to_gcp_map: dict) -> str | None:
+    """Convert a billing row to a dataset"""
+    try:
+        project_id = row['project']['id']
+    except TypeError:
+        project_id = None
+
+    topic = dataset_to_gcp_map.get(project_id, project_id)
+
+    # Default topic, any cost not clearly associated with a project will be considered
+    # overhead administrative costs. This category should be minimal
+    if not topic:
+        return DEFAULT_TOPIC
+
+    topic = RE_matcher.sub('', topic)
+    return topic
+
+
+def insert_new_rows_in_table(
+    table: str,
+    objs: list[dict[str, Any]],
+    dry_run: bool,
+    chunk_size=DEFAULT_BQ_CHUNK_SIZE,
+) -> int:
+    """Insert JSON rows into a BQ table"""
```

Comment:
Given how complicated this function is, maybe it deserves a slightly longer documentation string?

Reply:
Haha, yes absolutely :)